base/linalg/dense.jl

```diff
@@ -693,6 +690,309 @@ function inv(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(sin(Hermitian(A)))
+    end
+    return (exp!(im*A) - exp!(-im*A)) / 2im
+end
+
+"""
+    sincos(A::StridedMatrix)
+
+Compute the matrix sine and cosine of a square matrix `A`.
+
+# Examples
+```jldoctest
+julia> S, C = sincos(ones(2, 2));
+
+julia> S
+2×2 Array{Float64,2}:
+ 0.454649  0.454649
+ 0.454649  0.454649
+
+julia> C
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function sincos(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full.(sincos(Symmetric(A)))
+    end
+    c, s = reim(exp!(im*A))
+    return s, c
+end
+function sincos(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full.(sincos(Hermitian(A)))
+    end
+    X, Y = exp!(im*A), exp!(-im*A)
+    return (X - Y) / 2im, (X + Y) / 2
+end
+
+"""
+    tan(A::StridedMatrix)
+
+Compute the matrix tangent of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the tangent. Otherwise, the tangent is determined by calling [`exp`](@ref).
+
+# Examples
+```jldoctest
+julia> tan(ones(2, 2))
+2×2 Array{Float64,2}:
+ -1.09252  -1.09252
+ -1.09252  -1.09252
+```
+"""
+function tan(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full(tan(Symmetric(A)))
+    end
+    s, c = sincos(A)
+    return s / c
+end
+function tan(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full(tan(Hermitian(A)))
+    end
+    X, Y = exp!(im*A), exp!(-im*A)
+    return (X - Y) / (im * (X + Y))
+end
+
+"""
+    cosh(A::StridedMatrix)
+
+Compute the matrix hyperbolic cosine of a square matrix `A`.
+"""
+function cosh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(cosh(Hermitian(A)))
+    end
+    return (exp(A) + exp!(-A)) / 2
+end
+
+"""
+    sinh(A::StridedMatrix)
+
+Compute the matrix hyperbolic sine of a square matrix `A`.
+"""
+function sinh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(sinh(Hermitian(A)))
+    end
+    return (exp(A) - exp!(-A)) / 2
+end
+
+"""
+    tanh(A::StridedMatrix)
+
+Compute the matrix hyperbolic tangent of a square matrix `A`.
+"""
+function tanh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(tanh(Hermitian(A)))
+    end
+    X, Y = exp(A), exp!(-A)
+    return (X - Y) / (X + Y)
+end
+
+"""
+    acos(A::StridedMatrix)
+
+Compute the inverse matrix cosine of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the inverse cosine. Otherwise, the inverse cosine is determined by using
+[`log`](@ref) and [`sqrt`](@ref).
+
+# Examples
+```jldoctest
+julia> acos(cos(ones(2, 2)))
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function acos(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(acos(Hermitian(A)))
+    end
+    return -im * log(A + sqrt(A^2 - one(A)))
```

Comment:
`A^2 - I` is preferred for this kind of thing.

Note that this is suboptimal because it computes the Schur factorization twice.  As pointed out by [Aprahamian (2016)](eprints.ma.man.ac.uk/2432/01/covered/MIMS_ep2016_4.pdf), you can just compute the Schur factorization of `A` once.